contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,447 @@
+current target Object state, a controller may update the Object in place 
+provided that the associated version tracking information is updated as well.
+1. Otherwise, the controller must destroy the Object and recreate it from the 
+current target Object state.
+
+## Kubectl
+
+Modifications to kubectl to leverage controller history are an optional 
+extension. Users can trigger rolling updates and rollbacks by modifying their 
+manifests and using `kubectl apply`. Controllers will be able to detect 
+revisions to their target Object state and perform 
+[reconciliation](#target-object-state-reconciliation) as necessary.
+
+### Viewing History 
+As controllers can use strategic merge patching as a method for 
+[version equivalence](#version-equivalence) testing, so can clients use it as a 
+method for reconstructing specification type Objects at a particular revision. 
+We propose the following command.
+
+```bash
+> kubectl history <controller-kind> <controller-name>
+```
+
+In the above `controller-kind` indicates the Kind of the controller (e.g. 
+ReplicaSet, StatefulSet, DaemonSet) and `controller-name` indicates the `.Name` 
+of the request subject. We propose that the output of this command be a list 
+of the `.Names` of AnyStates that comprise the revision history of the 
+controller.
+
+In order to view the application of a particular revision to the controller we 
+propose the following additional flag.
+
+```bash
+> kubectl history <controller-kind> <controller-name> --revision <any-state-name>
+```
+
+The `--revsion` will GET the AnyState `.Named` `<any-state-name>` along with 
+controller. It will use strategic merge patch to merge the target Object state 
+contained in the AnyState with specification type Object in the controller's 
+manifest and to output the resulting manifest.
+
+- This output manifest represents the scale invariant application of a prior 
+target Object state to the current specification type Object.
+- This approach only works because of the encoding we propose for 
+[version snapshot construction](#version-snapshot-creation). Controllers that 
+use a different encoding will not be able to use this feature.
+
+### Rollback
+
+For future work, `kubeclt rollback` can be implemented in the general case as an 
+extension of the [above](#viewing-history ).
+
+```bash
+> kubectl rollback <controller-kind> <controller-name> --revision <any-state-name>
+```
+
+Here `kubectl rollback` simply applies the manifest generated by the 
+`kubectl history` command. We propose that the `--revision` argument be 
+optional, and that its exclusion indicates that the controller should be 
+reverted to its immediately prior state.
+
+## Tests
+
+1. Controllers can create an AnyState containing a revision of their target 
+Object state.
+1. Controllers can reconstruct their revision history.
+1. Controllers can't update an AnyState's `.Data`.
+1. Controllers can delete an AnyState to maintain their history with respect to 
```

Comment:
> can we set the revisionhistorylimit as 0 and hence the controllers dont generate AnyState at all ?

revisionhistorylimit=0 doesn't mean that the controllers are not generating history, eg. a DaemonSet in the middle of a rollout would need both the old and the new history object. revisionHistoryLimit should take effect only for history objects with no links (no running pods in the system generated by them).